implement GitOps CI/CD with ArgoCD

nadmax opened the pull request:
## What's new?

**Infra**
- Add Kubernetes manifests for backend and frontend deployment
-  Update deployment manifests to support image pull secrets
- Configure ArgoCD for automated GitOps deployments

**CI/CD**
- Add CD workflow with automated image builds, security scanning, and manifest updates
- Configure Docker layer caching for faster builds
- Integrate Trivy security scanning for vulnerability detection
- Use semantic versioning from git tags for image tagging

nadmax committed:
feat(infra): add Kubernetes deployment configuration with ArgoCD and GitHub Actions CI/CD

nadmax committed:
feat(infra): add frontend manifest for its deployment

nadmax committed:
feat(CI/CD): implement GitOps CI/CD pipeline with ArgoCD

- Add CD workflow with automated image builds, security scanning, and manifest updates
- Configure Docker layer caching for faster builds
- Integrate Trivy security scanning for vulnerability detection
- Use semantic versioning from git tags for image tagging
- Update deployment manifests to support image pull secrets
- Configure ArgoCD for automated GitOps deployments

nadmax committed:
chore(infra): update ingress host from 'terrascope.changeme.example' to 'terrascope.cloud''

nadmax committed:
chore(infra): create k8s namespace and move frontend ingress config

- Add new terrascope Kubernetes namespace
- Move frontend ingress configuration to k8s folder
- Delete ingress folder

sofianedjerbi commented on .github/workflows/cd.yml:
```diff
@@ -0,0 +1,161 @@
+            exit 0
+          fi
+
+          git commit -m "🚀 Deploy: Update images to ${{ needs.deploy.outputs.image-tag }} [skip ci]
```
imagine si le skip ci marche pas
je rigole... mais imagine quand meme (emoji feu)

nadmax replied:
https://docs.github.com/en/actions/how-tos/manage-workflow-runs/skip-workflow-runs
Impossible que ça ne fonctionne pas

sofianedjerbi replied:
d'ailleurs le smiley me donne envie de mourir + utilise le type `chore(release)` si tu aimes bien conventional commit (masochisme)

sofianedjerbi replied:
> https://docs.github.com/en/actions/how-tos/manage-workflow-runs/skip-workflow-runs Impossible que ça ne fonctionne pas

je rigole pitié

nadmax replied:
Je vais changer ça

sofianedjerbi commented on .github/workflows/cd.yml:
```diff
@@ -0,0 +1,161 @@
+          IMAGE_TAG="${{ needs.deploy.outputs.image-tag }}"
+          NEW_IMAGE="${{ env.REGISTRY }}/${{ env.IMAGE_OWNER }}/terrascope-backend:${IMAGE_TAG}"
+
+          sed -i "s|image: ${{ env.REGISTRY }}/${{ env.IMAGE_OWNER }}/terrascope-backend:.*|image: ${NEW_IMAGE}|g" \
```
ton approche est tres bien mais je partage ca parce que ca fait moins hacky. c'est pas une demande de changement tkt
```suggestion
          yq -i '.spec.template.spec.containers[0].image = strenv(NEW_IMAGE)' infra/k8s/deployments/backend.yml \
```

nadmax replied:
J'aime bien ta suggestion, je vais modifier c'est plus compréhensible ton snippet

sofianedjerbi commented on infra/k8s/frontend-ingress.yml:
tu preferes pas les httproute ?

nadmax replied:
C'est plus moderne que les ingress non?

sofianedjerbi replied:
Oui ! Mais il faut les CRDs de api gateway et un controller sur ton cluster. Ça te donne une bien meilleure séparation des responsabilités pour laisser des gens déployer en GitOps sur ton cluster sans tout péter

nadmax replied:
J'ai fait le remplacement

sofianedjerbi commented on .github/workflows/cd.yml:
```diff
@@ -0,0 +1,161 @@
+jobs:
+  deploy:
+    runs-on: ubuntu-latest
+    permissions:
```
il manque `security-events: write` pour sarif

nadmax replied:
Bien que je viens de l'ajouter, ça ne dit pas que ta review est outdated :shrug: 

*peut-être à cause du snippet sélectionné?*

sofianedjerbi reviewed (commented):
c'est mergeable pour moi ! je te laisse voir mes coms avant

nadmax committed:
chore(CD): change commit made by GitHub Actions

nadmax committed:
fix(CD): add missing security-events permission when deploying

nadmax committed:
chore(CD): remove unnecessary yq installation

nadmax committed:
chore(infra): replace Ingress with Gateway API HTTPRoute

sofianedjerbi commented on infra/k8s/gateway.yml:
Ca c'est coté cluster, l'app définit seulement la route

nadmax replied:
J'ai supprimé le gateway.yml, il est déjà déployé sur mon cluster

sofianedjerbi commented on infra/k8s/namespace.yml:
```diff
@@ -0,0 +1,4 @@
+apiVersion: v1
+kind: Namespace
```
Tu as déjà create namespace = true dans l'application, pourquoi le remettre?

nadmax replied:
Au temps pour moi, j'avais pas pris en compte qu'ArgoCD le crée au moment où j'ai crée cette conf

Je l'ai supprimé

sofianedjerbi commented on infra/argocd/application.yml:
```diff
@@ -0,0 +1,35 @@
+    path: infra/k8s
+  destination:
+    server: https://kubernetes.default.svc
+    namespace: terrascope
```
cool de l'avoir mis ici, tu peux l'enlever des deployments non ? il sera injecté direct par argocd (AFAIK)

nadmax replied:
J'ai retiré `namespace: terrascope` dans les deployments, les services et dans l'HTTPRoute

sofianedjerbi commented on infra/k8s/deployments/backend.yml:
```diff
@@ -0,0 +1,46 @@
+        - name: ghcr-secret
+      containers:
+        - name: terrascope-api
+          image: ghcr.io/lecafecloud/terrascope-backend:latest
```
pour eviter :latest tu peux utiliser argocd image updater . Latest passe quand meme pour le projet

nadmax replied:
J'ai ajouté des annotations dans `infra/argocd/application.yml` pour qu'ArgoCD Image Updater soit utiliser

nadmax replied:
J'ai aussi changé la config de la CD, parce que je n'ai plus besoin de mettre à jour les manifests, ArgoCD le fait automatiquement.
À la place c'est un job résumant le déploiement

sofianedjerbi reviewed (changes requested):
une fois que tu as mis gateway coté cluster tu peux merge no problem

nadmax committed:
chore(infra): migrate to ArgoCD Image Updater with semver strategy

- Add Image Updater annotations for automated deployments
- Remove update-k8s-manifests job from CI/CD pipeline
- Remove redundant namespace declarations from manifests
- Delete gateway.yml and namespace.yml
- Use semantic versioning (0.1.0) instead of SHA digests

Image Updater now automatically detects new versions and updates
deployments within 2 minutes, eliminating manual manifest updates.

nadmax committed:
chore(infra): add Kustomize for ArgoCD Image Updater

- Remove annotations in ArgoCD application config

- Remove deployment commit mention in notify-deploy job
